src/lib/operators/join_sort_merge.cpp

```diff
@@ -910,31 +891,73 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractJoinOperatorImpl {
 
     _perform_join();
 
-    // merge the pos lists into single pos lists
-    auto output_left = _concatenate_pos_lists(_output_pos_lists_left);
-    auto output_right = _concatenate_pos_lists(_output_pos_lists_right);
+    if (include_null_left || include_null_right) {
+      auto null_output_left = std::make_shared<PosList>();
+      auto null_output_right = std::make_shared<PosList>();
 
-    // Add the outer join rows which had a null value in their join column
-    if (include_null_left) {
-      for (auto row_id_left : *_null_rows_left) {
-        output_left->push_back(row_id_left);
-        output_right->push_back(NULL_ROW_ID);
+      // Add the outer join rows which had a null value in their join column
+      if (include_null_left) {
+        null_output_left->reserve(_null_rows_left->size());
+        null_output_right->insert(null_output_right->end(), _null_rows_left->size(), NULL_ROW_ID);
+        for (const auto& row_id_left : *_null_rows_left) {
+          null_output_left->push_back(row_id_left);
+        }
+      }
+      if (include_null_right) {
+        null_output_left->insert(null_output_left->end(), _null_rows_right->size(), NULL_ROW_ID);
+        null_output_right->reserve(_null_rows_right->size());
+        for (const auto& row_id_right : *_null_rows_right) {
+          null_output_right->push_back(row_id_right);
+        }
+      }
+
+      if (!null_output_left->empty()) {
+        _output_pos_lists_left.push_back(null_output_left);
+        _output_pos_lists_right.push_back(null_output_right);
```

Comment:
Probably a stupid question, but why don't we have to do the same check for the right side?

Reply:
The lists should have the same length. Added a DebugAssert.